ESM-only: Replace `esbuild-register` with hooks-api+`esbuild`'s `transform`-API

ndelangen opened the pull request:
Closes https://github.com/storybookjs/storybook/issues/31798

## What I did

This pull request introduces changes to improve module handling and transition toward ESM (ECMAScript Modules) compatibility. Key updates include modifications to package configurations, adjustments to utility functions, and updates to scripts for Storybook integration.

### Module Handling Updates:
* [`code/core/package.json`](diffhunk://#diff-1cf995d66081a270419b84c5b68236d9387d96d4dedf8223c6673df62eaa117cL175-R175): Replaced the `require` field with `default` to prioritize ESM compatibility for the `common` module.
* [`code/core/scripts/helpers/generatePackageJsonFile.ts`](diffhunk://#diff-cfa4570fd7bb941ad54da6d777107ebb14635fbe433b7c12a7ab4ab6be54d6e2R34-R40): Added a temporary hack to enforce ESM-only configuration for `common` module entries by setting the `default` field instead of `require`.

### Utility Function Adjustments:
* [`code/core/src/common/utils/interpret-require.ts`](diffhunk://#diff-18e05daf68a2d7dc08cf6b3665e6847606d2126741df5b180f1246ea70b8dffaL3-R21): Refactored the `interopRequireDefault` function to use dynamic `import` for ESM support, removing the dependency on `esbuild-register`. This simplifies the implementation and aligns with ESM standards.

### Storybook Integration Updates:
* [`code/package.json`](diffhunk://#diff-af9b72cb6dcf333969863361a126e03c67fb19ec94a062462e4d4bc3e2f801cbL48-R49): Updated the Storybook scripts to use `core/bin/index.cjs` instead of `node_modules/storybook/bin/index.cjs`, ensuring compatibility with the updated module structure.

## Checklist for Contributors

### Testing

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to communicate how to test your changes -->

#### The changes in this PR are covered in the following automated tests:

- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

<!-- Please include the steps to test your changes here. For example:

1. Run a sandbox for template, e.g. `yarn task --task sandbox --start-from auto --template react-vite/default-ts`
2. Open Storybook in your browser
3. Access X story

-->

### Documentation

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to indicate which documentation has been updated. -->

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli-storybook/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

  - `bug`: Internal changes that fixes incorrect behavior.
  - `maintenance`: User-facing maintenance tasks.
  - `dependencies`: Upgrading (sometimes downgrading) dependencies.
  - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
  - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
  - `documentation`: Documentation **only** changes. Will not show up in release changelog.
  - `feature request`: Introducing a new feature.
  - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
  - `other`: Changes that don't fit in the above categories.

   </details>

### 🦋 Canary release

<!-- CANARY_RELEASE_SECTION -->
This pull request has been released as version `0.0.0-pr-31803-sha-080533ec`. Try it out in a new sandbox by running `npx storybook@0.0.0-pr-31803-sha-080533ec sandbox` or in an existing project with `npx storybook@0.0.0-pr-31803-sha-080533ec upgrade`.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-31803-sha-080533ec`](https://npmjs.com/package/storybook/v/0.0.0-pr-31803-sha-080533ec) |
| **Triggered by** | @ndelangen |
| **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
| **Branch** | [`esm-only/add-node-loader-to-bin-replace-esbuild-register`](https://github.com/storybookjs/storybook/tree/esm-only/add-node-loader-to-bin-replace-esbuild-register) |
| **Commit** | [`080533ec`](https://github.com/storybookjs/storybook/commit/080533ec2fb81ff7258367c251917869cec83485) |
| **Datetime** | Thu Jun 19 11:46:30 UTC 2025 (`1750333590`) |
| **Workflow run** | [15757083337](https://github.com/storybookjs/storybook/actions/runs/15757083337) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=31803`_
</details>
<!-- CANARY_RELEASE_SECTION -->

<!-- BENCHMARK_SECTION -->
<!-- BENCHMARK_SECTION -->


<!-- greptile_comment -->

## Greptile Summary

Major architectural change replacing esbuild-register with native ESM module handling and esbuild's transform API across the Storybook codebase.

- Added new `loader.mjs` using Node.js hooks API and esbuild transform for TypeScript/ESM compatibility
- Converted key utilities like `interopRequireDefault` to use dynamic `import()` instead of require
- Updated package.json files to prioritize ESM by replacing `require` fields with `default` export conditions
- Migrated module resolution from `require.resolve` to `import.meta.resolve` + `fileURLToPath` across core server components
- Added proper async/await handling for module loading in core functions like `loadCustomPresets` and `loadCustomWebpackConfig`



<!-- /greptile_comment -->

ndelangen committed:
Refactor: Update Storybook commands to use core/bin and enhance ESM support

- Changed Storybook UI commands in package.json to reference core/bin instead of node_modules/storybook.
- Updated package.json in core to use 'default' for common module exports.
- Introduced hooks.mjs and loader.mjs for improved ESM handling and TypeScript support.
- Modified interopRequireDefault function to handle dynamic imports and fallback to CommonJS.

ndelangen committed:
Refactor: Update custom preset and webpack config loading to support async operations

- Changed `loadCustomPresets` and `loadCustomWebpackConfig` functions to be asynchronous, allowing for better handling of dynamic imports.
- Updated the usage of these functions in the relevant files to await their results, ensuring proper configuration loading.
- Removed deprecated `esbuild-register` dependency from package.json.

ndelangen committed:
Update dependencies and improve async support in Storybook configuration

- Bump Node.js version requirement to 22.16.0 in package.json.
- Update .nvmrc to reflect the new Node.js version.
- Add esbuild dependency version range to package.json and yarn.lock.
- Modify getStorybookInfo calls to be asynchronous in relevant files for better handling of dynamic imports.
- Introduce new hooks.mjs and loader.mjs for enhanced ESM support and TypeScript handling.

ndelangen committed:
Merge branch 'next' into esm-only/add-node-loader-to-bin-replace-esbuild-register

ndelangen committed:
Update yarn.lock to replace 'empathic' with 'find-up' dependency

ndelangen committed:
Add ESM support for TypeScript in Storybook bin

- Introduced hooks.mjs and loader.mjs for enhanced ESM handling and TypeScript transformation.
- Updated index.cjs to utilize the new hooks for loading TypeScript files.
- Included bin/**/* in package.json to ensure proper file inclusion.

ndelangen committed:
Update esbuild dependency version range in package.json and yarn.lock

ndelangen committed:
Fix loader resolution in sandbox-parts.ts to use import.meta for ESM compatibility

JReinhold commented on code/core/bin/loader.mjs:
```diff
@@ -0,0 +1,25 @@
+
+export function load(url, context, nextLoad) {
+  /** Convert TS to ESM using esbuild */
+  if (url.match(/\..?tsx?$/)) {
```
https://jsbenchmark.com/#eyJjYXNlcyI6W3siaWQiOiJhS1Q4TkpQYVJiQUFaQ2l2bnFIU1giLCJjb2RlIjoiREFUQS5mb3JFYWNoKChuYW1lKSA9PiBuYW1lLmVuZHNXaXRoKCcudHMnKVxuICAgICAgICAgICAgIHx8IG5hbWUuZW5kc1dpdGgoJy5tdHMnKVxuICAgICAgICAgICAgIHx8IG5hbWUuZW5kc1dpdGgoJy5jdHMnKVxuICAgICAgICAgICAgIHx8IG5hbWUuZW5kc1dpdGgoJy50c3gnKVxuICAgICAgICAgICAgIHx8IG5hbWUuZW5kc1dpdGgoJy5tdHN4JylcbiAgICAgICAgICAgICB8fCBuYW1lLmVuZHNXaXRoKCcuY3RzeCcpXG4gICAgICAgICAgICApIiwibmFtZSI6ImVuZHNXaXRoIiwiZGVwZW5kZW5jaWVzIjpbXX0seyJpZCI6ImxiR0h3aHRxREN5aXlvVWdpQlVqMCIsImNvZGUiOiJEQVRBLmZvckVhY2goKG5hbWUpID0-IC9cXC5bbWNdP3RzeD8kL2kudGVzdChuYW1lKSkiLCJuYW1lIjoicmVnZXgiLCJkZXBlbmRlbmNpZXMiOltdfV0sImNvbmZpZyI6eyJuYW1lIjoiQmFzaWMgZXhhbXBsZSIsInBhcmFsbGVsIjp0cnVlLCJnbG9iYWxUZXN0Q29uZmlnIjp7ImRlcGVuZGVuY2llcyI6W119LCJkYXRhQ29kZSI6InJldHVybiBbLi4uQXJyYXkoMTAwMCkua2V5cygpLm1hcChTdHJpbmcpXSJ9fQ

ndelangen committed:
Refactor: Enhance ESM support and clean up bin files

- Added `./bin/loader.mjs` export to package.json for improved module resolution.
- Removed deprecated `hooks.mjs` files from bin directories.
- Updated `index.cjs` to directly require the dist files, simplifying the loading process.
- Adjusted `generatePackageJsonFile` to handle new export formats for core-server presets.
- Replaced `require.resolve` with `import.meta.resolve` in multiple files for better ESM compatibility.

ndelangen committed:
Update tests to reflect changes in presets and build index functionality

- Enhanced the `preset` object in `presets.test.ts` to include detailed properties for better testing coverage.
- Marked the `should resolve all addons & presets in correct order` test as skipped in `presets.test.ts` to prevent execution during current test runs.
- Updated `build-index.test.ts` to skip the `should build index` test, indicating a temporary change in test execution status.

ndelangen committed:
Refactor: Introduce defaultFavicon constant and update tests for consistency

- Added `defaultFavicon` constant to `constants.ts` for better favicon management.
- Updated `common-preset.ts` to import `defaultFavicon`.
- Modified tests in `favicon.test.ts`, `stories-json.test.ts`, and `StoryIndexGenerator.test.ts` to mock `defaultFavicon` and `defaultStaticDirs` for consistent test behavior.

ndelangen committed:
Refactor: Improve ESM compatibility by using fileURLToPath

- Updated `standalone.ts` and `common-preset.ts` to replace `import.meta.resolve` with `fileURLToPath` for better ESM support.
- Ensured correct directory resolution for package.json files in both modules.

ndelangen committed:
Refactor: Enhance ESM compatibility in build-static.ts

- Updated `build-static.ts` to use `fileURLToPath` for resolving the directory of `package.json`, improving ESM support.
- This change aligns with recent updates to other files for consistent handling of module paths.

ndelangen committed:
Refactor: Update package.json and generatePackageJsonFile for ESM compatibility

- Changed the export in `package.json` from `require` to `default` for telemetry.
- Enhanced the `generatePackageJsonFile` function to include additional checks for ESM-only compatibility, specifically for telemetry files.

ndelangen committed:
Refactor: Update loader resolution in sandbox-parts.ts for ESM compatibility

- Changed the loader resolution in `sandbox-parts.ts` from `import.meta.resolve` to a direct string reference for `@storybook/addon-docs/mdx-loader`, improving consistency with recent ESM updates.

ndelangen committed:
Refactor: Remove esbuild version range from package.json and yarn.lock

- Eliminated the esbuild version range from `package.json` files in `cli-storybook` and `create-storybook`, as well as from `yarn.lock`, to streamline dependency management and avoid potential conflicts.

ndelangen committed:
Refactor: Standardize quotes in CircleCI configuration

- Updated the CircleCI configuration file to use consistent double quotes for strings, enhancing readability and maintaining uniformity across the configuration.

ndelangen committed:
Refactor: Enhance loader function to support additional TypeScript file extensions

- Updated the loader function in `loader.mjs` to recognize and process multiple TypeScript file extensions, including `.mts`, `.cts`, `.mtsx`, and `.ctsx`, improving compatibility with various TypeScript module formats.

ndelangen committed:
Update extra development dependencies in sandbox-parts.ts

- Changed the version of `@storybook/test-runner` in `sandbox-parts.ts` from `0.22.1--canary.d4862d0.0` to `0.23.1--canary.fd3c654.0` to incorporate the latest features and fixes.

ndelangen committed:
Fix typo in extra development dependencies in sandbox-parts.ts

- Corrected the package name from `storybook/test-runner` to `@storybook/test-runner` in `sandbox-parts.ts` to ensure proper dependency resolution.

ndelangen committed:
Merge branch 'next' into esm-only/add-node-loader-to-bin-replace-esbuild-register

ndelangen committed:
Update extra development dependency version in sandbox-parts.ts

- Changed the version of `@storybook/test-runner` from `0.23.1--canary.fd3c654.0` to `0.23.1--canary.b6bd5f6.0` in `sandbox-parts.ts` to incorporate the latest updates.

ndelangen committed:
Add Node.js 22.15.0 installation step in CircleCI configuration

- Added a new step in the CircleCI configuration to install Node.js version 22.15.0 using NVM, ensuring the correct Node version is used during the build process.

ndelangen committed:
Update CircleCI configuration for Node.js installation

- Modified the Node.js installation step in the CircleCI config to use wget instead of curl, and added commands to set up NVM environment variables for better compatibility and reliability during builds.

ndelangen committed:
Update Node.js version in CircleCI configuration to 22.16.0

- Changed the Node.js installation step in the CircleCI config to use version 22.16.0, ensuring the build process utilizes the latest features and improvements.

ndelangen committed:
Update minification options and extra dev dependency version

- Set `minifyWhitespace` to `false` in `bundle.ts` to prevent whitespace minification during the build process.
- Updated the version of `@storybook/test-runner` in `sandbox-parts.ts` from `0.23.1--canary.b6bd5f6.0` to `0.23.1--canary.2094e4b.0` to incorporate the latest updates.

ndelangen committed:
Update extra development dependency version in sandbox-parts.ts

- Updated the version of `@storybook/test-runner` in `sandbox-parts.ts` from `0.23.1--canary.2094e4b.0` to `0.23.1--canary.49024ff.0` to incorporate the latest updates.

ndelangen committed:
Update extra development dependency version in sandbox-parts.ts

- Updated the version of `@storybook/test-runner` in `sandbox-parts.ts` from `0.23.1--canary.49024ff.0` to `0.23.1--canary.de6531a.0` to incorporate the latest updates.

ndelangen committed:
Merge branch 'sb10/esm-only' into esm-only/add-node-loader-to-bin-replace-esbuild-register

ndelangen committed:
cache-bust

ndelangen committed:
Import frameworkPackages from get-storybook-info utility in baseGenerator.ts

ndelangen committed:
Merge branch 'sb10/esm-only' into esm-only/add-node-loader-to-bin-replace-esbuild-register

ndelangen committed:
Refactor require registration and update minification options

- Added a check to ensure the 'storybook/bin/loader.mjs' is registered only once in interpret-require.ts.
- Changed the minification option `minifyWhitespace` to `true` in bundle.ts to enable whitespace minification during the build process.

ndelangen committed:
Merge branch 'esm-only/add-node-loader-to-bin-replace-esbuild-register' of https://github.com/storybookjs/storybook into esm-only/add-node-loader-to-bin-replace-esbuild-register

ndelangen committed:
add log for testing windows

ndelangen committed:
more debugging

ndelangen committed:
attempt at fixing windows

ndelangen committed:
actually use the resolved

ndelangen committed:
Update path resolution for Windows in interpret-require.ts

- Added a check for the Windows platform to ensure file paths are correctly resolved to file URLs when they do not start with 'file:'. This addresses issues with module resolution on Windows systems.

ndelangen committed:
Refactor presets in presets.test.ts to simplify preset structure

- Removed unnecessary properties from the preset object in the loadPreset tests, streamlining the test cases for better clarity and maintainability.

ndelangen committed:
Remove console log statement in interopRequireDefault function to clean up code

ndelangen committed:
Update gitHead references in package.json files across multiple addons and frameworks to maintain consistency and reflect the latest commit.
